File: src/freight_fate/audio_loops.py
"""Reusable sustain-loop-with-release playback for BASS (sound_lib) streams.

A real "attack -> sustain -> release" sound -- a horn held down, a siren, an
engine that idles then spins down -- should loop only a short interior region
while it is held, then play its natural release tail once let go. Plain
whole-file looping instead replays the attack every cycle and never lets the
release ring out.

This module is the BASS half of that model. :class:`SustainLoop` drives a raw
``BASS_ChannelSetSync`` position sync that seeks the stream back to the loop
start each time playback reaches the loop end; :meth:`SustainLoop.release`
removes the sync so playback flows past the loop end through to the end of the
file. Because the sync is a *mixtime* sync, the seek happens during mixing and
the loop is seamless.

Nothing here depends on the game's :class:`~freight_fate.audio.AudioEngine`; it
operates on any ``sound_lib`` stream exposing ``.handle``, ``get_frequency``,
and ``seconds_to_bytes``, so it can be reused by any BASS-backed sound. The
pygame fallback backend has no equivalent seek/sync primitive and implements
its own segment-based version.

Loop points may be given in samples or seconds; :func:`to_seconds` is the
shared conversion so every caller agrees on units.
"""

from __future__ import annotations

import logging

__all__ = ["to_seconds", "SustainLoop"]

log = logging.getLogger(__name__)


def to_seconds(pos: float, units: str = "samples", freq: float | None = None) -> float:
    """Convert a loop point to seconds.

    ``units`` is ``"samples"`` (``freq`` in Hz is then required) or
    ``"seconds"`` (``pos`` is returned unchanged).
    """
    if units == "seconds":
        return float(pos)
    if units == "samples":
        if not freq:
            raise ValueError("a stream frequency is required to convert samples to seconds")
        return float(pos) / float(freq)
    raise ValueError(f"unknown loop-point units: {units!r}")


class SustainLoop:
    """A live sustain loop on one BASS stream, seeking start<-end each cycle.

    Construct it on a stream that is created **non-looping** and already
    playing (or about to play): it installs a mixtime position sync at
    ``loop_end`` that seeks back to ``loop_start``. Call :meth:`release` to let
    the release tail play out, or :meth:`stop` when tearing the loop down.
    """

    def __init__(self, stream, loop_start: float, loop_end: float, *, units: str = "samples"):
        # Imported lazily so environments without BASS (pygame-only) can still
        # import this module for ``to_seconds`` without pulling in sound_lib.
        from sound_lib.external.pybass import (
            BASS_POS_BYTE,
            BASS_SYNC_MIXTIME,
            BASS_SYNC_POS,
            SYNCPROC,
            BASS_ChannelRemoveSync,
            BASS_ChannelSetPosition,
            BASS_ChannelSetSync,
        )

        self._stream = stream
        self._handle = stream.handle
        self._remove_sync = BASS_ChannelRemoveSync
        self._pos_byte = BASS_POS_BYTE
        self._released = False
        self._sync = 0

        try:
            freq = stream.get_frequency()
        except Exception:  # pragma: no cover - defensive; freq only used for samples
            freq = None
        start_s = to_seconds(loop_start, units, freq)
        end_s = to_seconds(loop_end, units, freq)
        if end_s <= start_s:
            raise ValueError("loop_end must be after loop_start")
        self._start_byte = int(stream.seconds_to_bytes(start_s))
        end_byte = int(stream.seconds_to_bytes(end_s))

        # BASS keeps a raw pointer to this callback for as long as the sync is
        # registered; pin it to the instance so Python does not free it out
        # from under the mixer thread. The seek is done at mixtime so it lands
        # exactly on the loop boundary with no audible gap.
        def _on_loop_end(handle, channel, data, user):
            BASS_ChannelSetPosition(channel, self._start_byte, self._pos_byte)

        self._proc = SYNCPROC(_on_loop_end)
        self._sync = BASS_ChannelSetSync(
            self._handle, BASS_SYNC_POS | BASS_SYNC_MIXTIME, end_byte, self._proc, None
        )

    @property
    def released(self) -> bool:
        return self._released

    def release(self) -> None:
        """Remove the loop sync so playback continues into the release tail.

        Idempotent: safe to call more than once (e.g. a defensive stop after a
        key release).
        """
        if self._released:
            return
        self._released = True
        try:
            self._remove_sync(self._handle, self._sync)
        except Exception:  # pragma: no cover - stream may already be gone
            log.debug("SustainLoop.release: sync already removed", exc_info=True)

    def stop(self) -> None:
        """Tear the loop down. The caller is responsible for stopping the stream."""
        self.release()

File: crates/ff-core/src/audio_loops.rs
//! The backend-free half of sustain-loop-with-release playback.
//!
//! A real "attack -> sustain -> release" sound -- a horn held down, a siren,
//! an engine that idles then spins down -- should loop only a short interior
//! region while it is held, then play its natural release tail once let go.
//! Plain whole-file looping instead replays the attack every cycle and never
//! lets the release ring out.
//!
//! The Python module (`freight_fate/audio_loops.py`) carried both halves: the
//! unit maths here, and a `SustainLoop` class that installs a BASS *mixtime*
//! position sync seeking the stream back to the loop start each time playback
//! reaches the loop end, and removes it on `release()` so playback flows past
//! the loop end through to the end of the file. That BASS half needs a live
//! stream handle and the `SYNCPROC` trampoline kept alive for the mixer
//! thread, so it lives in the game crate: `freight_fate::audio` wires a
//! [`SustainLoopSpec`] onto a stream (`start_sustain_loop`, the horn's only
//! caller) and tears it down (`release_sustain_loop`). What stays here is
//! everything that can be checked without a device: the unit conversion every
//! caller must agree on, and the loop-point validation the BASS class did in
//! its constructor.
//!
//! Port of `freight_fate/audio_loops.py` (pure parts).

use std::fmt;

/// Loop-point units: samples (needs a stream frequency) or seconds.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LoopUnits {
    Samples,
    Seconds,
}

impl LoopUnits {
    /// The Python keyword the call sites pass (`units="samples"`).
    pub fn parse(units: &str) -> Option<Self> {
        match units {
            "samples" => Some(Self::Samples),
            "seconds" => Some(Self::Seconds),
            _ => None,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Samples => "samples",
            Self::Seconds => "seconds",
        }
    }
}

/// The loop-point errors the Python code raised as `ValueError`.
#[derive(Debug, Clone, PartialEq)]
pub enum LoopPointError {
    /// `to_seconds` was asked to convert samples with no (or zero) frequency.
    MissingFrequency,
    /// `units` was neither `"samples"` nor `"seconds"`.
    UnknownUnits(String),
    /// `loop_end` is not after `loop_start`.
    InvertedPoints,
}

impl fmt::Display for LoopPointError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::MissingFrequency => {
                write!(
                    f,
                    "a stream frequency is required to convert samples to seconds"
                )
            }
            Self::UnknownUnits(units) => write!(f, "unknown loop-point units: {units:?}"),
            Self::InvertedPoints => write!(f, "loop_end must be after loop_start"),
        }
    }
}

impl std::error::Error for LoopPointError {}

/// Convert a loop point to seconds.
///
/// `units` is `"samples"` (`freq` in Hz is then required) or `"seconds"`
/// (`pos` is returned unchanged).
pub fn to_seconds(pos: f64, units: &str, freq: Option<f64>) -> Result<f64, LoopPointError> {
    match LoopUnits::parse(units) {
        Some(units) => to_seconds_units(pos, units, freq),
        None => Err(LoopPointError::UnknownUnits(units.to_string())),
    }
}

/// [`to_seconds`] with the units already parsed.
pub fn to_seconds_units(
    pos: f64,
    units: LoopUnits,
    freq: Option<f64>,
) -> Result<f64, LoopPointError> {
    match units {
        LoopUnits::Seconds => Ok(pos),
        LoopUnits::Samples => match freq {
            // Python's `if not freq` treats 0.0 and None alike.
            Some(freq) if freq != 0.0 => Ok(pos / freq),
            _ => Err(LoopPointError::MissingFrequency),
        },
    }
}

/// A sustain loop described in the caller's units: the loop region the
/// backend must seek over while the sound is held.
///
/// The horn is `SustainLoopSpec::samples(11816, 12379)` (`HORN_LOOP_START`
/// / `HORN_LOOP_END` in the audio module). [`SustainLoopSpec::resolve`] is
/// what the Python `SustainLoop.__init__` did before touching BASS: convert
/// both points to seconds and refuse an inverted region.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SustainLoopSpec {
    pub loop_start: f64,
    pub loop_end: f64,
    pub units: LoopUnits,
}

/// Loop points in seconds, ready for the backend's own byte conversion.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ResolvedSustainLoop {
    pub start_s: f64,
    pub end_s: f64,
}

impl SustainLoopSpec {
    pub const fn samples(loop_start: f64, loop_end: f64) -> Self {
        Self {
            loop_start,
            loop_end,
            units: LoopUnits::Samples,
        }
    }

    pub const fn seconds(loop_start: f64, loop_end: f64) -> Self {
        Self {
            loop_start,
            loop_end,
            units: LoopUnits::Seconds,
        }
    }

    /// Both points in seconds for a stream at `freq` Hz (needed for sample
    /// units only), refusing a region whose end is not after its start.
    pub fn resolve(&self, freq: Option<f64>) -> Result<ResolvedSustainLoop, LoopPointError> {
        let start_s = to_seconds_units(self.loop_start, self.units, freq)?;
        let end_s = to_seconds_units(self.loop_end, self.units, freq)?;
        if end_s <= start_s {
            return Err(LoopPointError::InvertedPoints);
        }
        Ok(ResolvedSustainLoop { start_s, end_s })
    }
}

#[cfg(test)]
mod tests {
    //! Unit-level checks for the reusable sustain-loop helper.
    use super::*;

    // The shipped horn's loop points (audio.HORN_LOOP_START / HORN_LOOP_END).
    const HORN_LOOP_START: f64 = 11816.0;
    const HORN_LOOP_END: f64 = 12379.0;

    #[test]
    fn test_to_seconds_samples_uses_frequency() {
        assert_eq!(to_seconds(44100.0, "samples", Some(44100.0)).unwrap(), 1.0);
        let got = to_seconds(11816.0, "samples", Some(44100.0)).unwrap();
        assert!((got - 11816.0 / 44100.0).abs() < 1e-12);
    }

    #[test]
    fn test_to_seconds_seconds_passthrough_ignores_frequency() {
        assert_eq!(to_seconds(0.5, "seconds", Some(1.0)).unwrap(), 0.5);
        assert_eq!(to_seconds(0.5, "seconds", None).unwrap(), 0.5);
    }

    #[test]
    fn test_to_seconds_rejects_samples_without_frequency() {
        assert_eq!(
            to_seconds(11816.0, "samples", None),
            Err(LoopPointError::MissingFrequency)
        );
        assert_eq!(
            to_seconds(11816.0, "samples", Some(0.0)),
            Err(LoopPointError::MissingFrequency)
        );
    }

    #[test]
    fn test_to_seconds_rejects_unknown_units() {
        assert_eq!(
            to_seconds(1.0, "frames", Some(44100.0)),
            Err(LoopPointError::UnknownUnits("frames".into()))
        );
    }

    // The three BASS-stream tests of the Python file
    // (`test_sustain_loop_computes_byte_positions_from_samples`,
    // `test_sustain_loop_release_is_idempotent`,
    // `test_sustain_loop_rejects_inverted_points`) need a live stream; the
    // first two belong to the game crate's audio tests. The byte arithmetic
    // and the inversion check they exercise are pinned here without a device.

    #[test]
    fn test_sustain_loop_computes_byte_positions_from_samples() {
        let spec = SustainLoopSpec::samples(HORN_LOOP_START, HORN_LOOP_END);
        let resolved = spec.resolve(Some(44100.0)).unwrap();
        // Verified against the shipped 44100 Hz horn asset: 16-bit stereo is
        // four bytes a frame, so the loop start lands on byte 47264.
        let start_frame = resolved.start_s * 44100.0;
        assert!((start_frame - HORN_LOOP_START).abs() < 1e-9);
        assert_eq!(start_frame.round() as u64 * 4, 47264);
    }

    #[test]
    fn test_sustain_loop_rejects_inverted_points() {
        let spec = SustainLoopSpec::samples(HORN_LOOP_END, HORN_LOOP_START);
        assert_eq!(
            spec.resolve(Some(44100.0)),
            Err(LoopPointError::InvertedPoints)
        );
    }
}
